acceptance: Refactors java acceptance tests

rohangulati opened the pull request:
Fixes #4591

This moves the java acceptance test to separate java files inside `acceptance/java` package to ease editing. Positive java test cases live under `acceptance/java/success` and negative test cases under `acceptance/java/fail`. The go test code invokes a bash scripts `run_java_test.sh` which compiles and runs `TestRunner.java` inside the test docker container. `TestRunner.java` is responsible for running all java test under `acceptance/java/success` and `acceptance/java/fail`. It does so by accepting 2 command line arguments `[expectedOutcome]` `[test_dir]`. `TestRunner` also makes sure that all generated files (.class and .pk8) files are cleaned up.


<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10659)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=10659) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=10659) before we can accept your contribution.<br/><hr/>**Rohan Gulati** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

tamird commented:
CI failure looks related to some issues on master (should go away with a rebase).

tamird commented:
While you're rebasing, please include the entire PR description in your commit message.

rohangulati committed:
Refactors java acceptance test

Fixes #4591

This moves the java acceptance test to separate java files inside `acceptance/java` package to ease editing. Positive java test cases live under `acceptance/java/success` and negative test cases under `acceptance/java/fail`. The go test code invokes a bash scripts 'run_java_test.sh' which compiles and runs 'TestRunner.java` inside the test docker container. 'TestRunner.java` is responsible for running all java test under 'acceptance/java/success` and 'acceptance/java/fail'. It does so by accepting 2 command line arguments [expectedOutcome] [test_dir]. 'TestRunner' also makes sure that all generated files (.class and .pk8) files are cleaned up.

rohangulati commented:
@tamird Rebase with master and included the description in the commit message

knz commented:
Thanks for your help. Some minor comments left.

---

Reviewed 3 of 6 files at r1.
Review status: 3 of 6 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/java/TestRunner.java, line 31 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWXgBTFj2WIWH2RFhLx:-KWXgBTGTNVA1c0RWfVs:b-y7yqoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/TestRunner.java#L31)):*
> ```Java
>     public static void main(String[] args) throws Exception {
>         if (args.length < 2) {
>             throw new Exception("unexpected : missing arguments. usage : java TestRunner [SUCCESS/FAIL] [dir]");
> ```

In English there is no space between a colon (:) and the previous word. e.g. `"unexpected: missing ..."`  (also see below)
Put `"usage: ..."` on a different line; e.g. `"unexpected: missing arguments\nusage: java ..."`

---

*[pkg/acceptance/java/TestRunner.java, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWXgZ38DDKulHS71e86:-KWXgZ38DDKulHS71e87:b-7bf2ls) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/TestRunner.java#L37)):*
> ```Java
>         ExpectedOutcome expectedOutcome = ExpectedOutcome.parse(mode);
>         if (expectedOutcome == null) {
>             throw new Exception("unexpected : missing arguments. usage : java TestRunner [SUCCESS/FAIL] [dir]");
> ```

The null value doesn't say the argument was missing, but rather that it wasn't correct. Make the exception more informative.

---

*[pkg/acceptance/java/TestRunner.java, line 88 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWXgtwi1XZ-VA_F5Syn:-KWXgtwjtuhy0QilMQvO:bmyy73w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/TestRunner.java#L88)):*
> ```Java
> 
>         for (File file : files) {
>             if (!file.getName().contains(".java")) {
> ```

Simplify: `if (file.getName().contains(".hava")) { javaFiles.add(file); }`

also, use `endsWith` not `contains`.

---

*[pkg/acceptance/java/TestRunner.java, line 179 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWXhFvzQ0jwguBZfQJi:-KWXhFvzQ0jwguBZfQJj:b-de60hc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/TestRunner.java#L179)):*
> ```Java
>         }
>     }
> }
> ```

Missing newline at end of file.

---

*[pkg/acceptance/java/fail/JavaFailureTest.java, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWXhYndC7MVBz0UjDpf:-KWXhYndC7MVBz0UjDpg:b-rbhfvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/fail/JavaFailureTest.java#L46)):*
> ```Java
>             throw new Exception("unexpected: DROP TABLE reports " + res + " rows changed, expecting 0");
>         }
>         stmt = conn.prepareStatement("SELECT 1, 2 > ?, ?::int, ?::string, ?::string, ?::string, ?::string, ?::string");
> ```

Move this test into a separate file.

---

*[pkg/acceptance/java/fail/JavaFailureTest.java, line 67 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWXh_dfAugRlk-8WpP1:-KWXh_dfAugRlk-8WpP2:b-br7u90) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/fail/JavaFailureTest.java#L67)):*
> ```Java
>             throw new Exception("unexpected");
>         }
>         stmt = conn.prepareStatement("CREATE TABLE accounts (id INT PRIMARY KEY, balance INT, cdate DATE)");
> ```

Move this test (this CREATE together with the INSERT below) into a separate file.

---

*[pkg/acceptance/java/success/JavaSuccessTest.java, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWXhiaV3R4z504_7Aoj:-KWXhiaV3R4z504_7Aok:b-j6e1zf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/success/JavaSuccessTest.java#L48)):*
> ```Java
>             throw new Exception("unexpected: DROP TABLE reports " + res + " rows changed, expecting 0");
>         }
>         stmt = conn.prepareStatement("SELECT 1, 2 > ?, ?::int, ?::string, ?::string, ?::string, ?::string, ?::string");
> ```

Split test here (see comment above)

---

*[pkg/acceptance/java/success/JavaSuccessTest.java, line 69 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWXhkzuwMH5LCB8MT4H:-KWXhkzuwMH5LCB8MT4I:b-j6e1zf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/success/JavaSuccessTest.java#L69)):*
> ```Java
>             throw new Exception("unexpected");
>         }
>         stmt = conn.prepareStatement("CREATE TABLE accounts (id INT PRIMARY KEY, balance INT, cdate DATE)");
> ```

Split test here (see comment above)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-:-KWXhnuOUh47UkHdtAVa:b35w90b)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 6 files at r1.
Review status: 4 of 6 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/java/run_cleanup.sh, line 4 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWYXRpPNosu-MwVVLLS:-KWYXRpPNosu-MwVVLLT:b-nzt6bj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/run_cleanup.sh#L4)):*
> ```Shell
> 
> # cleans all {.class, .pk8} files in the current directory
> find . -type f -name "*.class" -exec rm -f {} \;
> ```

```
find . -type f -name '*.class' -or -name '*.pk8' -exec rm -f {} \;
```

---

*[pkg/acceptance/java/run_java_test.sh, line 2 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWYXczr3WorsV8rSHJG:-KWYXczr3WorsV8rSHJH:bfu9j1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/run_java_test.sh#L2)):*
> ```Shell
> #!/usr/bin/env bash
> set -e
> ```

`set -euo pipefail`

---

*[pkg/acceptance/java/run_java_test.sh, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWYXlHVRKzAhcReftZ8:-KWYXlHVRKzAhcReftZ9:b-yjgb85) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/run_java_test.sh#L9)):*
> ```Shell
> export CLASSPATH=$CLASSPATH:$JAVA_ROOT/success:$JAVA_ROOT/fail:$JAVA_ROOT
> 
> cd $JAVA_ROOT
> ```

how come this is necessary?

---

*[pkg/acceptance/java/TestRunner.java, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWYYtRaVbhIjtXf9yIh:-KWYYtRaVbhIjtXf9yIi:b-19cs2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/TestRunner.java#L9)):*
> ```Java
> 
> /*
> This class compiles and runs all java files under a directory.
> ```

perhaps i'm missing something, but why is this test runner written in java? it seems to me that this thing amounts to little more than shelling out to various tools, and that there is little or no benefit to it being written in java.

given that java is not a core competency for us, I would prefer to see this written in Go.

---

*[pkg/acceptance/java/TestRunner.java, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWYYehs43hpyhsz1EqB:-KWYYehs43hpyhsz1EqC:bbl3t0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/TestRunner.java#L59)):*
> ```Java
>                         break;
>                     }
>                     // This test should have "PASSED" but "FAILED
> ```

broken sentence.

---

*[pkg/acceptance/java/fail/JavaFailureTest.java, line 7 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWYY0TS8t4ECG9srXlm:-KWYY0TS8t4ECG9srXln:bb4o6kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/fail/JavaFailureTest.java#L7)):*
> ```Java
>     public static void main(String[] args) throws Exception {
>         Class.forName("org.postgresql.Driver");
>         String DB_URL = "jdbc:postgresql://";
> ```

this connection string construction code is duplicated; can it be shared?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659)*
<!-- Sent from Reviewable.io -->

rohangulati commented:
Review status: 4 of 6 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/java/TestRunner.java, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659#-KWYYtRaVbhIjtXf9yIh:-KWYesu56xDs_7yjbGtg:b-iwpc1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a2760159936e90ff558963493952be7d29119c6/pkg/acceptance/java/TestRunner.java#L9)):*
> Not specific reason. It was as per the suggestion "the different Java tests should be executed one after another by test logic written in Java" mention in the issue(#4578) description. I agree that this should be written in go. I will rewrite this in go.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10659)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@knz, @dt, @mjibson what remains to be done here?

dt commented:
At the very least we need to rebase it, in particular to pick up 66c28b. 

Beyond that, looks like there's @tamird's objection to the loop that runs the individual tests being on the java side rather than in Go, but that potentially could get changed later if/when it becomes a pain point.

spencerkimball commented:
@dt could we get this merged, assuming it's straightforward? And update / close #4591.

justinj commented:
I'm going to close this since it appears to have died and #4591 was closed by #19321.
